### DISCUSSION TITLE
Is hot reload for user and password creds available?

### DISCUSSION THREAD

**Discussion by msberyanov** [etcd-io/jetcd#1377] [Q&A]
Hi @lburgazzoli. 

We have a task to hot reload creds to etcd (they can be changed from vault). Is it allowed in your library? How to process situations where watches on key are already set and currently listening? Should I recreate client every time creds change? How would you do this?

Thank you.

**Comment by msberyanov** [comment]
I guess it is fine to change **user**, **password** fields in _ClientConnectionManager_ through builder() and then _AuthCredential_  will update token with new info?

**Reply by lburgazzoli** [reply]
It could work but I won't assume that changing a builder value then get property handled. We would need to add a proper way like providing a proper authentication provider/supplier

if you have time to work on such a feature, I'd be happy to help/review

**Reply by msberyanov** [reply]
Thank you! Tell me please what to do?

**Reply by msberyanov** [reply]
@lburgazzoli ?

**Reply by lburgazzoli** [reply]
As today we have a `ClientBuilder.user` and `ClientBuilder.password` methods, so I guess we should introduce a new interface, something like:

```java
interface AuthCredentialProvider {
    ByteSequence username();
    ByteSequence password();
}
```

So you can plug it with your vault based solution. But mind that, this is the first thing I have on top of my mind.

**Reply by msberyanov** [reply]
https://github.com/IBM/etcd-java/pull/101 I have done it in the neighbouring lib. Most likely it is either done in yours, or should be done. It is working.

**Reply by lburgazzoli** [reply]
as I said, it should be done if you want to have something working reliably
